### PR TITLE
drivers: add lis2mdl mag driver

### PR DESCRIPTION
**Pull request by bkueng**
It has less functionality than the lis3mdl, but supposedly the lis2mdl is better than the lis3mdl.

### CODE REVIEW

File: src/drivers/magnetometer/lis2mdl/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,43 @@
+############################################################################
+#
+#   Copyright (c) 2016 PX4 Development Team. All rights reserved.
```
Comment (dagar) [inline src/drivers/magnetometer/lis2mdl/CMakeLists.txt:3]: year

File: src/drivers/drv_sensor.h [hunk 2]
```diff
@@ -160,6 +160,8 @@
 #define DRV_MAG_DEVTYPE_UAVCAN	0x88
 #define DRV_DIST_DEVTYPE_UAVCAN	0x89
 
+#define DRV_MAG_DEVTYPE_LIS2MDL  0xb0
```
Comment (dagar) [inline src/drivers/drv_sensor.h:163]: Maybe `0x0C` grouped with most of the other mags?

File: src/drivers/magnetometer/lis2mdl/lis2mdl.cpp [hunk 3]
```diff
@@ -0,0 +1,202 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2012-2015 PX4 Development Team. All rights reserved.
```
Comment (dagar) [inline src/drivers/magnetometer/lis2mdl/lis2mdl.cpp:3]: year

File: src/drivers/magnetometer/lis2mdl/lis2mdl.cpp [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+	perf_free(_sample_perf);
+	perf_free(_comms_errors);
+	perf_free(_range_errors);
+	perf_free(_conf_errors);
+}
+
+int
+LIS2MDL::measure()
+{
+#pragma pack(push, 1)
```
Comment (dagar) [inline src/drivers/magnetometer/lis2mdl/lis2mdl.cpp:72]: This doesn't need to be packed, then `-Wno-cast-align` can be removed from CMakeLists.txt.

File: src/drivers/magnetometer/lis2mdl/lis2mdl.cpp [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+{
+	I2CSPIDriverBase::print_status();
+	perf_print_counter(_sample_perf);
+	perf_print_counter(_comms_errors);
+	PX4_INFO("poll interval:  %u", _measure_interval);
+	_px4_mag.print_status();
+}
+
+int
+LIS2MDL::reset()
```
Comment (dagar) [inline src/drivers/magnetometer/lis2mdl/lis2mdl.cpp:176]: Delete?

File: src/drivers/magnetometer/lis2mdl/lis2mdl.h [hunk 6]
```diff
@@ -0,0 +1,163 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2015 PX4 Development Team. All rights reserved.
```
Comment (dagar) [inline src/drivers/magnetometer/lis2mdl/lis2mdl.h:3]: year

File: src/drivers/magnetometer/lis2mdl/lis2mdl_i2c.cpp [hunk 7]
```diff
@@ -0,0 +1,148 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2013-2015 PX4 Development Team. All rights reserved.
```
Comment (dagar) [inline src/drivers/magnetometer/lis2mdl/lis2mdl_i2c.cpp:3]: year

File: src/drivers/magnetometer/lis2mdl/lis2mdl_i2c.cpp [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+LIS2MDL_I2C::LIS2MDL_I2C(int bus, int bus_frequency) :
+	I2C("LIS2MDL_I2C", nullptr, bus, LIS2MDLL_ADDRESS, bus_frequency)
+{
+	_device_id.devid_s.devtype = DRV_MAG_DEVTYPE_LIS2MDL;
+}
+
+int
+LIS2MDL_I2C::ioctl(unsigned operation, unsigned &arg)
```
Comment (dagar) [inline src/drivers/magnetometer/lis2mdl/lis2mdl_i2c.cpp:91]: The IOCTLs can be removed.

File: src/drivers/magnetometer/lis2mdl/lis2mdl_i2c.cpp [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+LIS2MDL_I2C_interface(int bus, int bus_frequency);
+
+device::Device *
+LIS2MDL_I2C_interface(int bus, int bus_frequency)
+{
+	return new LIS2MDL_I2C(bus, bus_frequency);
+}
+
+LIS2MDL_I2C::LIS2MDL_I2C(int bus, int bus_frequency) :
+	I2C("LIS2MDL_I2C", nullptr, bus, LIS2MDLL_ADDRESS, bus_frequency)
```
Comment (dagar) [inline src/drivers/magnetometer/lis2mdl/lis2mdl_i2c.cpp:85]:
```suggestion
	I2C(DRV_MAG_DEVTYPE_LIS2MDL, "LIS2MDL_I2C", bus, LIS2MDLL_ADDRESS, bus_frequency)
```

File: src/drivers/magnetometer/lis2mdl/lis2mdl_i2c.cpp [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+{
+	_device_id.devid_s.devtype = DRV_MAG_DEVTYPE_LIS2MDL;
```
Comment (dagar) [inline src/drivers/magnetometer/lis2mdl/lis2mdl_i2c.cpp:87]:
```suggestion
```

File: src/drivers/magnetometer/lis2mdl/lis2mdl_i2c.cpp [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+	case DEVIOCGDEVICEID:
+		return CDev::ioctl(nullptr, operation, arg);
+
+	default:
+		return  -EINVAL;
+	}
+}
+
+int
+LIS2MDL_I2C::probe()
```
Comment (dagar) [inline src/drivers/magnetometer/lis2mdl/lis2mdl_i2c.cpp:107]: Note for later, this is an example of the pattern we need to get right to cleanup the multi-interface driver mess. The probe shouldn't be specific to the interface.

File: src/drivers/magnetometer/lis2mdl/lis2mdl_main.cpp [hunk 12]
```diff
@@ -0,0 +1,142 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2012-2015 PX4 Development Team. All rights reserved.
```
Comment (dagar) [inline src/drivers/magnetometer/lis2mdl/lis2mdl_main.cpp:3]: year

File: src/drivers/magnetometer/lis2mdl/lis2mdl_spi.cpp [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+LIS2MDL_SPI_interface(int bus, uint32_t devid, int bus_frequency, spi_mode_e spi_mode);
+
+device::Device *
+LIS2MDL_SPI_interface(int bus, uint32_t devid, int bus_frequency, spi_mode_e spi_mode)
+{
+	return new LIS2MDL_SPI(bus, devid, bus_frequency, spi_mode);
+}
+
+LIS2MDL_SPI::LIS2MDL_SPI(int bus, uint32_t devid, int bus_frequency, spi_mode_e spi_mode) :
+	SPI("LIS2MDL_SPI", nullptr, bus, devid, spi_mode, bus_frequency)
```
Comment (dagar) [inline src/drivers/magnetometer/lis2mdl/lis2mdl_spi.cpp:85]:
```suggestion
	SPI(DRV_MAG_DEVTYPE_LIS2MDL, "LIS2MDL_SPI", bus, devid, spi_mode, bus_frequency)
```

File: src/drivers/magnetometer/lis2mdl/lis2mdl_spi.cpp [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+{
+	_device_id.devid_s.devtype = DRV_MAG_DEVTYPE_LIS2MDL;
```
Comment (dagar) [inline src/drivers/magnetometer/lis2mdl/lis2mdl_spi.cpp:87]:
```suggestion
```

File: src/drivers/magnetometer/lis2mdl/lis2mdl_spi.cpp [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+	if (data != ID_WHO_AM_I) {
+		DEVICE_DEBUG("LIS2MDL bad ID: %02x", data);
+		return -EIO;
+	}
+
+	return OK;
+}
+
+int
+LIS2MDL_SPI::ioctl(unsigned operation, unsigned &arg)
```
Comment (dagar) [inline src/drivers/magnetometer/lis2mdl/lis2mdl_spi.cpp:118]: Can be deleted.

**Comment by bkueng** [comment]
Lot's of copy-paste. I did another cleanup round.

**Comment by dagar** [comment]
I suppose we should actually build it? 

https://github.com/PX4/Firmware/blob/3a7fd3f63f6a9dbdb93f8695edf30cb82b99350a/src/drivers/magnetometer/CMakeLists.txt#L40

**Comment by bkueng** [comment]
> I suppose we should actually build it?

I wasn't adding it for flash, and as no other board than v5x is using it.

**Comment by ryankierteal** [comment]
Has this been tested in hardware?  I'm thinking specifically of SPI on the STM32F7.  By default the LIS2MDL SPI interface operates in 3-wire mode (p. 25 of datasheet):  
image
I haven't found anywhere that bit 2 of CFG_REG_C is being set, and it doesn't look like 3-wire SPI is supported by the driver framework.  

Perhaps `LIS2MDL_SPI::init()` needs this bit set before `ADDR_WHO_AM_I` is read, and `LIS2MDL::init()` needs to read `ADDR_CFG_REG_C`, OR the readback result with `CFG_REG_C_BDU`, then write the new value to `ADDR_CFG_REG_C`?

**Comment by bkueng** [comment]
I only tested this with I2C. Feel free to make the required changes for SPI and create a PR.

**Pull request merged**